### DISCUSSION TITLE
Standalone video player?

### DISCUSSION THREAD

**Discussion by FrayxRulez** [ffmpeginteropx/FFmpegInteropX#352] [General]
While this library is quite cool `MediaPlayer(Element)` definitely isn't.
Have you considered building your own?

**Comment by brabebhin** [comment]
That would be pretty huge undertaking that neither me nor @lukasf  have the time for. Technically it is possible to create our own wrapper around the MediaFoundation interfaces that MediaPlayer wraps, but even that is quite complicated.

What exactly do you not like about MediaPlayer? It is mostly quite OK, except for the bugs in MediaPlaybackList that can be usually worked around with some effort.

**Reply by FrayxRulez** [reply]
Oh yeah, no, I mean, I would like not to use MF at all.
MediaPlayer is extremely slow and adds a lot of latency when playing videos.
As well, I'm quite convinced that the threading model used is quite wrong as the UI often hangs while changing player source.

**Reply by FrayxRulez** [reply]
So I would like a quite barebone player control that given a source can render the video and audio without all the overhead generated by MediaPlayer. What I don't like, specifically, is all the time needed from SetSource to the first frame being displayed on screen.
Having gapless playback for videos would also be great.

**Reply by brabebhin** [reply]
You do have gapless playback for video, MediaPlaybackList works great with videos. Actually, MediaPlaybackList is quite an amazing feat of engineering all things considered. It does have bugs -> you can''t use video effects with the MediaPlaybackList, you have to choose if you want gapless videos or video effects. 

The threading model of MediaPlayer shouldn't impede the UI, all events are async on background threads. I'd argue it is still better than say, android's implementation of media engines which throw all events on the UI thread. I use message queue to throttle computations when media changes. This significantly improves performance. Most of the samples Microsoft has on media playback and handling the events in UI are barebones and not quite suited for a production application.

Are you observing this latency between SetSource and when playback starts with our lib or you are using Microsoft's decoders?

**Reply by FrayxRulez** [reply]
I experience the latency with either your library or Microsoft's decoders. 
Videos that I'm trying to play gapless are h265 and there's a big gap in between (especially considering that the videos themselves aren't longer than 5 seconds)

**Reply by brabebhin** [reply]
I'd say this is simply your computer not having the power to do what it takes. H265 is difficult to deal with especially at high frame resolutions. Only the top GPUs of relatively new generations can handle that (I think nvidia 3000 series or newer can do it). Probably you are observing gaps because the first file finishes rendering too fast and the 2nd hasn't had the chance to buffer yet.

A SSD also helps.

**Reply by FrayxRulez** [reply]
Well, it's a Razer Blade 2023 with 64 GB of RAM, a GeForce RTX 4070 and an i9-13950HX, if this is not enough power I'm not sure how videos can even play on a normal user device? 🤣

Videos are just 720p, 30 fps. Can't even imagine going higher than this. Still, I always had latency from SetSource to the first frame appearing on screen from the very first version of the app (released 7 years ago) regardless from the video/device. This is one of the most frequent feedback I receive when talking about media in the app.

(Want to underline that now I'm testing in a empty project with ~20 lines of code in total, just in case the next point would be that it is my code being wrong)

**Reply by FrayxRulez** [reply]
(I'm not saying that the problem is FFmpegInteropX either, just in case 😄 I come in peace)

**Reply by brabebhin** [reply]
If you got some samples that you can share I could take a look and see what happens on my PC
Unfortunately latency depends on a lot of things, some of them outside of MediaFoundation's control.

According to this review, the dGPU does have latency problems:

https://www.notebookcheck.net/Razer-Blade-16-Early-2023-RTX-4090-Review-Core-i9-13950HX-beast-with-world-s-first-dual-mode-mini-LED-UHD-display.702200.0.html

You can use the same software to test for system latency on your own individual device and see what's up.

**Reply by FrayxRulez** [reply]
Here's one:
https://github.com/ffmpeginteropx/FFmpegInteropX/assets/8152849/21303213-9eed-413a-98b5-14b0f84f2886

It takes ~2 seconds from SetSource to the first frame regardless of software/hardware decoder, MF/ffmpeg when using `MediaPlayer`.
If I open the same with libVLC it will play pretty much instantly, and multiple ones will go gapless as well.

**Reply by brabebhin** [reply]
To be fair, we don't have anywhere near the resources of VLC so we can't really compare to them. Nor do we even try, although we are sometimes successful in outperforming them.

**Reply by FrayxRulez** [reply]
Yeah, once again my complain is about Microsoft code and not yours :)

**Reply by brabebhin** [reply]
Can i get a second file as well?

I was referring to creating our own rendering pipeline, in addition to the current decoding one.

I'm also working on a transcoding pipeline. eta next year lol.

**Reply by lukasf** [reply]
I cannot reproduce your issue here. The gap between files after setting source is few milliseconds, not nearly a second or even two. No major difference to any other file. I suggest you try a GPU driver update and see if you can resolve the issue.

For true gapless you need the MediaPlaybackList. But even without, there should not be much of a delay, unless there is something wrong with the system.

**Reply by FrayxRulez** [reply]
One of the most common complains that I receive about the media part of the app (both internal and external feedback) is the fact that media player is slow to start the playback. This applied to every single laptop I tried the app on over the last 7 years, either before or after adding FFmpegInteropX.

**Reply by brabebhin** [reply]
I tried it with my own app, playback is also almost instant indeed.
Let's try something else: do our own samples also exhibit the same latency when you open the files? What about the native Windows Media Player (if you are on windows 11) or the movies+TV app (on windows 10 and 11)? Those 2 also use MediaFoundation.

You should always keep the AutoPlay property true if you want your MediaPlayer to be snapy. If you use AutoPlay = false and manually control when playback starts, segmentation of execution may also introduce latency.

If our samples do exhibit the same latency, you can attach a performance profiler using VS and see what eating away at your performance.

**Reply by FrayxRulez** [reply]
Of course that the same issue can be reproduced with Windows Media Player and Movies+TV, as well as your samples. I wonder if at this point the issue here is a different perception of the same thing? As is: for you waiting 2 seconds for the first video to play and then few milliseconds to a second of blank screen in between videos is something acceptable while for me it isn't?

**Reply by brabebhin** [reply]
The only 2 second wait I can think of is the cold start of the app. If the app is already running, then playback starts instantly, even if the app hasn't been playing anything. But I do not believe you are referring to a cold start when you say this.

Have you tried testing your system with LatencyMon? I do believe that you have a 2 second lag, but I don't think MediaFoundation is automatically to blame. You are in a somewhat special position in being able to actually reproduce what your users are reporting.

Is your app accessible? Would it be possible for me to download it?

**Reply by lukasf** [reply]
My player based on FFmpegInteropX + MediaPlayerElement is also super fast in opening files. Playback starts near-instant. There is this slight black flicker when switching files, which is unfortunate. I'd wish we could configure MediaPlayerElement to keep displaying the last image while switching files, like VLC does. But it is no big deal really. Most other players show that behavior as well (and most of them are even a bit slower).

One thing I always do to improve file switching times is calling OpenAsync on the MediaSource before setting it, like this:

```
var ffmpegSource = await FFmpegMediaSource.CreateFromStreamAsync(readStream, Config);
var playbackItem = FFmpegMSS.CreateMediaPlaybackItem();
await playbackItem.Source.OpenAsync();
mediaPlayer.Source = playbackItem;
```

This will do part of the work before the actual source switching and will lead to shorter flicker and faster percieved switching.

I am still not 100% sure about your actual scenario. Do the 2 seconds include app startup, or is this the time taken to open a file after the app already started? Is this really 2 seconds or is it just that you feel it should be faster? About the gap when you switch to a different file: Is this while one is currently playing, or is it when reaching file end and switching to next file? On file end you can get gapless using MediaPlaybackList (be aware of some issues with that). Or you can at least reduce gaps by pre-opening the next file ahead of time (be sure to also call Source.OpenAsync as shown above). If you only ever begin opening the next file after the previous ended, the gap will of course be longer.

To clarify things, best would be if you could do a screen capture showing your scenarios, using the same file you uploaded. Then it would be a lot clearer to us what you see, and we could say if this seems normal to us or not. You can do that with the Windows 11 Snipping Tool, Xbox Gamebar or Geforce Experience. Or a lot of other tools.

@brabebhin By the way, I noticed that the WinUI sample app has noticeably shorter gap between files, compared to CPP and CS samples!

**Reply by brabebhin** [reply]
haha, I thought it was me imagining that lol.
I think MediaPlayerElement does have some changes in winUI and it isn't simply a wrapped control from UWP. Another source of performance is quicker file access. The winRT streaming APIs are somewhat faster in winUI. StorageFile still sux though.

Overall I am quite pleased with the performance of winUI. I ported my player and it is lightning fast now: cold startup is almost as good in debug mode as UWP is in release mode. File access is so much smoother. The only downside is that you have to cook the full screen support by yourself.

**Reply by FrayxRulez** [reply]
The app can be downloaded from [here](https://www.microsoft.com/store/productId/9N97ZCKPD60Q).
The feature that would require the faster video player is still WIP.
It's a bit hard to provide a screen recording of the test app because the clips that I'm using are not recorded by me and I'm not sure if who recorded them is happy to have them shared over GitHub 😅 but I'll try to give you something.

What surprises me here is that I've always considered the MediaPlayer(Element?) slow, regardless from the app, device and video played.
Now I can agree on the fact that on some specific file/device scenarios this is even more noticeable--but this still doesn't solve my problem: users will still complain with me as other softwares on their machine (apart from Media Player that I feel not so many people use) can smoothly reproduce videos without long bufferings and hangs.

**Reply by brabebhin** [reply]
Haha, I actually have that app installed xD

**Reply by brabebhin** [reply]
ok I see now what you mean.
The delay isn't exactly 2 seconds but it is indeed noticeable and I'm on battery saver. My guess is that the problem comes from the way you store your files. 
What framework is your app based on? is it pure UWP or winUI? is it some cross platform wrapper?

Ok If i plug in it is almost instant as well, but still not as fast as my app or the samples. Is there some kind of preview playing?

**Reply by lukasf** [reply]
Wow this is an impressive app! Very well done.
I will consider switching from official Telegram client to Unigram!

Since the app looks really elaborate, I will take the time for an elaborate answer. I have tested this on my 5 years old laptop, to have something more representable than my high-end dev machine. Indeed I saw some issues with video performance.

Here are my findings and recommendations:
- When opening a feed with videos in them, most of them start playing instantly, with a delay close to zero. However, many videos (especially high-res or demanding codec) play choppy. Looking at the task manager, I see that the app does not use hardware decoding for any of these, although the formats are supported! You can easily play multiple files with HW acceleration at the same time. Why do you use software decoding?
- If I play the same videos in our sample app, even with forced software decoding, the same files play 100% smooth. Do you use any special settings on the Config class? Do you modify the MediaStreamSource somehow, like setting BufferTime or Live mode or something like that? Try to remove any special settings you have. Performance should be same as in our sample apps.
- Only when opening a file in fullscreen view, GPU decoding is used.
- When opening the feed, a play button is shortly shown over the video and then disappearing after playback started. This is unneccessary and it will definitely increase the percieved loading delay! You should try to never show any kind of loading indicators, if load times are expected to be shorter than a second. The important thing here is the perceived performance. Showing things even for a fraction of a second during load will increase percieved load times. You should already know which files you will auto-play. Do not show a play button on them. This will make the app feel faster.
- When I click on a file to play full screen, the first time there is a delay. Probably not even a second, but noticeable. When I close it and open it again, there is still a delay, but much shorter, more in the milliseconds range. Still, the load time on 2nd and further loads seem long to me. Do you already start loading the file on Click or do you load it once the animation completed? I cannot say for sure, but it feels like the video is only really opened when the animation ended.
- When I use the arrows to switch between files in fullscreen mode, the load times are a lot longer! Also the transport controls flicker *twice*. I am pretty sure that there is something really wrong there. My feeling is that for some reason, the video file is opened twice, or source is assigned twice, causing double delay. Twice the load times, twice flicker. I know that some XAML controls like to raise SelectionChanged events twice. Especially with the FlipView, I had some serious issues like that, and I had to use workarounds to avoid calling some expensive operations twice. I think in some situations, FlipView even set SelectedItem to null and then back to the actual value. You should definitely look into that and make sure that the file is really opened only once and the source is only set once. Always set source in code, so you can put breakpoints and check when it is called.
- Something I would strongly recommend: When in fullscreen mode, why don't you load the previous and next video ahead of time? You could open the file, create a MediaPlayer (AutoPlay=false) and assign the source. When the user switches to the file, you only need to call Play() and there will be zero load delay. Percieved performance will be vastly improved!
- Flicker is always annoying. I would consider hiding the transport controls on file change and showing them only after playback started.
- When playing files fullscreen, I see that all videos in the background feed are still playing. So while you open and play a file fullscreen, you have potentially one, two or even more very demanding files being CPU decoded and rendered in the background. Even though they are not even visible (or barely visible). This will unnceccessarily eat resouces (CPU and also GPU for rendering) and make your load times suffer. I'd recommend you pause all videos in a feed as soon as the fullscreen viewer is opened. Even if HW acceleration is enabled for the feed videos, it's a waste of resources. You can immediately continue them when the fullscreen player animates out.
- You might also consider this: You can assign one MediaPlayer to multiple MediaPlayerElements at the same time. If you play a video in the feed with HW decoding and click on it, you could theoretically assign the same MediaPlayer instance to the MediaPlayerElement of the fullscreen viewer. This would allow 100% seamless switching to fullscreen mode without any load delay. Only issue I saw with that: You cannot use the stock play or pause buttons on the transport controls. But calling Play() or Pause() on the MediaPlayer works without issues. You can easily replace the play/pause button on the transport controls with custom buttons which raise events, and then you hook to the events and control playback directly on the MediaPlayer. This is probably a bit tricky to get right, still I wanted to mention the possibility.
- If you want to reduce first time delay, you could try opening one or more short dummy files in the background and assign them to a MediaPlayer, and then cleanup as soon as MediaOpened comes. Not sure how much this really helps, but it will sure load all required MF dlls and initialize some media stuff.

**Reply by brabebhin** [reply]
On top of what lukasf suggested, you can also have multiple MediaPlayers getting ready to play files. Swapping them around will increase the responsiveness of the app.

You play one file, you can already start preparing the next in a completely separated MediaPlayer instance and simply assign that to your Element when it is ready to play. The second MediaPlayer can prepare on a background thread to reduce stuttering. 

For such chaining of files, MediaPlaybackList will also work great and reduce your lag by a lot. You can dynamically add or remove files from it as you want.

**Reply by lukasf** [reply]
Yes, when I talk about preloading video, I also ment not only creating the FFmpegMediaSource, but also creating a MediaPlayer and assigning the source, so the MediaPlayer does already initialize and build the MF FilterGraph. Assigning an initialized MediaPlayer to MediaPlayerElement is super fast. The delays you see come from MediaPlayer initialization. Just make sure to also cleanup properly when videos move out of view (properly dispose everything, including the MediaPlayer).

I think MediaPlaybackList is not the right thing for your app. It only makes sense when you show a single video viewer and you want the next video to immediately begin as soon as the previous ended, having gapless playback inside a player window. This is not how your app works. Users have to click buttons or touch/swipe to start the next video, and a new player window will move into view and start playing. Preloading previous/next videos into a MediaPlayer, ready to start, is what will greatly improve the user experience when navigating through video files.

@brabebhin By the way, the shorter gap in WinUI was only because the WinUI sample I had installed was from a different branch than the CPP/CS samples. On master branch, the samples do set mediaPlayer.Source = null before even starting to open a new file. On the auto-ref branch I had removed that to allow more seamless switching. When I build the CS/CPP samples from that branch, I do not see any speed difference to WinUI.

**Reply by FrayxRulez** [reply]
Sorry for the delay (actually I can’t find you on Telegram, but as well I go with the same username) and glad to hear that you like the app. 

Thank you for the very detailed reply, I’ll try to elaborate further: when the videos are playing in the chat history (the feed) there’s no MediaPlayer nor MF involved: we have a custom player that is used by animated stickers as well that renders frames using ffmpeg (“raw” one, not your library) directly onto a WriteableBitmap PixelBuffer that is then invalidated on screen using CompositionTarget.Rendering. After many years of experiments (the implementation was rewritten ~5 times and we passed through MediaElement, MediaPlayerElement, D2D1, etc) this ended up being the best solution as it’s extremely light in resources usage and doesn’t require any kind of initialization at all. MediaPlayer just can’t keep it up when you have to load and unload possibly hundreds of videos very fast. 
The gallery however does use MediaPlayerElement, even if at the moment it’s a single Element and a single Player (as it was built in times where MediaPlayerElement didn’t exist yet and it was recommended not to use multiple MediaElements at the same time). The app does NOT use FFmpegInteropX yet there, so what you see is just a standard MediaSource consuming a custom IRandomAccessStream that takes care of reading the file from the disk while the app downloads the file from the web (we don’t use HTTP). 
The view itself is quite custom, I rarely use built in controls because they tend to be quite buggy, so no FlipView involved. (The app is UWP)

**Reply by brabebhin** [reply]
For the real time chat, i think having multiple MediaPlayers handling each individual video, with AutoPlay false should be quicker.

You need a virtual listview here. The preloading should happen far in advance. Not sure if the default control can help you.

**Reply by FrayxRulez** [reply]
I have absolutely no plans to use multiple MediaPlayers in the chat history :) that would be very far from being optimized in terms of resources usage and general app performances. MediaPlayer isn’t really made for this kind of scenarios. The problem isn’t even the time needed for initialization here, but all the overhead added by such a control that would significantly decrease scrolling performances.

**Reply by brabebhin** [reply]
MediaPlayer is not a control.

**Reply by FrayxRulez** [reply]
This doesn’t really make any difference, in a way or another you have to display it on the screen :) the point stays the same.

**Reply by brabebhin** [reply]
Well, the MediaPlayer has a frame server mode, which allows you to render video to any Direct3D surface (that includes image controls). You won;t get subtitles this way, but it is still a very efficient way to do things, and it is slightly less overhead than MediaPlayerElement. 

As a reference, I did use a custom MediaPlayerElement and a MediaPlayer in frame server mode to obtain some functionality that wasn't available from the default control. I'd render the frames to the Placeholder album art image inside the MediaPlayerElement. This would prevent stuff like having a blank screen when changing files (it would display the last frame of the previous file while loading the next). And video effects that would go hand in hand with gapless video playback. And saving the exact frame that was being displayed. All the while the MediaPlayerElement would still display the subtitles for me :)

**Reply by lukasf** [reply]
You say you used the custom solution with WriteableBitmap because its low resource usage. I do not agree with that. Both video files I tested did not play properly in the feed. CPU usage was very high. One of the video was stuttering and not fluid. The other video was even worse, it only rendered a frame every few seconds. Sometimes I was not even sure if it is playing. This happened on my laptop which was pretty high-end 5 years ago, probably a standard PC these days. Both videos play flawless and 100% fluid using FFmpegMediaSource with HW decoding at very low resource usage. Even when using SW decoding, both files are playing at the same time, nice and fluid using FFmpegMediaSource. CPU usage is similar to your solution, but the playback is smooth and nice. If your users complain that the app cannot play videos smooth, then I bet this is the reason. Your solution might be fast to start and low on RAM (which most PCs have more than enough), but you burn a lot of other resources instead (CPU and battery - GPU decoding is much more efficient!), and the result is not even fluid on a lot of systems. It might work well for low-res stuff. But these days, FullHD is pretty standard and lots of people start recording 4K. Codecs are more and more demanding. Good luck SW decoding multiple 4K files. Many PCs won't even be able to SW decode a single 4K stream properly.

Of course, you need to limit the number of videos you have prepared at the same time. And you need to properly cleanup the ones not needed anymore. I would not play more than 2 or 3 files at the same time. And maybe keep 2-3 more prepared. You could also have a pool of MediaPlayers, so you re-use those instead of creating new ones all the time.

The MediaPlayerElement is actually pretty lightweight, if you do not display the transport controls. And you do not need to connect a MediaPlayer to it until the point where you want to start playback. Rendering is fully hardware accelerated and should not have any real impact on scrolling or animations. The only thing that is slow is resizing a MediaPlayerElement (true resize, not transform) while it is playing, because it will allocate new textures on each size change. But video sizes in your feed are constant, so this won't be an issue for you.

Frankly, I never tried using MediaPlayerElement in a way like in your feed. But I am sure you will continue to get complaints about performance, if you do not use HW decoding in the feed. I also do not understand why your software decoding is so much slower than our SW decoder. Maybe you should try using our FrameGrabber, calling ExtractNextVideoFrameAsync in a loop. It uses the same underlying decoders. It even allows to grab frames into an existing IBuffer, so you could pass the WriteableBitmap's buffer right into it (that is exactly the scenario why that overload was added). FFmpeg will use that buffer as decoder render target, so there is zero copying involved (copying large SW frames in RAM is slow!). Still, I'd rather recommend going for HW decoding, since that is the only way to get truly smooth playback across a vast amount of target devices.

**Reply by brabebhin** [reply]
It is also worth mentioning syncing between audio and video streams, which is also very important to proper playback.

I'd go for the preload solution with multiple MediaPlayers. Limit your preload to 5-10 videos and recycle the MediaPlayer objects. It should only cost you a few dozen MB at most. Having multiple MediaPlayers is actually a supported scenario and I've used it successfully in the past.

Remember, RAM is cheap and windows is pretty good at RAM management. When optimising, it is almost always better to optimize CPU and GPU usage at the cost of higher RAM usage.

**Reply by FrayxRulez** [reply]
I only receive complains about the full screen gallery so far. Videos rarely play automatically in the feed, it’s mostly GIFs and stickers that are quite low resolution MP4/WEBM files. Of course you won’t use more than 2-3 MediaPlayerElements at the same time, but a single message can contain a hundred custom emojis and even ignoring these you can have dozens of GIFs visible at the same time on screen by just opening the menu to send GIFs. 

Worth mentioning that we did receive a lot of complains about performances when using MediaPlayerElement in the feed.

**Reply by brabebhin** [reply]
I honestly don't know what else to advise you. I definitely think MF is fast enough for your goal (after all, the vast majority of Windows applications, messaging apps included use MF at some point or another).

As for standalone players, you don't really have much choice: libVLC does provide its own MediaPlayerElement for Xamarin Forms. Technically FFmpeg can also render video to any HWND. Pretty sure various game engines will also provide basic video playback capabilities. But all of these come with both advantages and disadvantages. Creating a standalone player is not quite the scope of this library, although I did give it some thought at some point. Too much work.

**Reply by lukasf** [reply]
For the gallery, I have already listed how things can be improved. Especially the load times when switching files within the gallery can be reduced to effectively zero by preloading files into ready-to-play MediaPlayer instances. This should be pretty simple. As opposed to the feed, you do not need to load hundreds of files super fast. You only need to keep 3 files opened, and files cannot change faster than the user can click or swipe. No worries abour resources here.

For the load time from clicking on an item in the feed to the player playing: When I open the video in our sample app and press Ctrl+Return to re-open it, the time till playback starts is shorter than the duration of the opening animation you use. So with this solution, you should be able to get a video ready to play when the animation ended, without any visible delay. Please note that this is when re-using one (or more) existing MediaPlayers. When I create a new MediaPlayer, the total delay is definitely longer! So you really want to use a pool of MediaPlayers where you fetch one for the next job and put unused ones back after resetting their source.

And I verified with a modified sample app: If I open a file in one MediaPlayer without even playing it, and then open the file again in a new MediaPlayer and render it, I do not see the long "first video playback after app startup" load delay. So opening a hidden dummy file at app startup should indeed help you get rid of the first time playback delay. And then you could even put that player into the pool, since it is already initialized and will be able to start playback faster than a new instance.

As I see it, all the complain points in the gallery should be solveable, to make it feel like there is more or less zero delay.

Creating a custom player is definitely out of scope here. It is way too much work to get right for our small team, including frame synced video playback, sync audio playback and subtitle rendering...

**Comment by FrayxRulez** [comment]
In the end I opted for forking libVLC and libVLCSharp to dynamically link to ffmpeg and remove any other unneeded modules. This way it's quite compact in size and performance (even with a single player, no preloading, etc) is pretty much up to the expectations.

So if you're not planning anything like this, I'll just close this.
